### DISCUSSION TITLE
Inconsistent value returned for `siglen` from EVP_DigestSign for EC

### DISCUSSION THREAD

**Discussion by svenkata9** [openssl/openssl#23180] [Q&A] [answered]
The first time when I call `EVP_DigestSign` will `NULL` for `sig`, it returns the size of the memory to be allocated for the signature via `siglen`. This value sometimes is more than the value returned from the same call with valid memory allocated for `sig` with `siglen`. Why is this discrepancy?

```
static int ecdsa_sign(void *vctx, unsigned char *sig, size_t *siglen,
                      size_t sigsize, const unsigned char *tbs, size_t tbslen)
{
    PROV_ECDSA_CTX *ctx = (PROV_ECDSA_CTX *)vctx;
    int ret;
    unsigned int sltmp;
    size_t ecsize = ECDSA_size(ctx->ec);

    if (!ossl_prov_is_running())
        return 0;

    if (sig == NULL) {
        *siglen = ecsize;
        return 1;
    }

#if !defined(OPENSSL_NO_ACVP_TESTS)
    if (ctx->kattest && !ECDSA_sign_setup(ctx->ec, NULL, &ctx->kinv, &ctx->r))
        return 0;
#endif

    if (sigsize < (size_t)ecsize)
        return 0;

    if (ctx->mdsize != 0 && tbslen != ctx->mdsize)
        return 0;

    if (ctx->nonce_type != 0) {
        ret = ossl_ecdsa_deterministic_sign(tbs, tbslen, sig, &sltmp,
                                            ctx->ec, ctx->nonce_type,
                                            ctx->mdname,
                                            ctx->libctx, ctx->propq);
    } else {
        ret = ECDSA_sign_ex(0, tbs, tbslen, sig, &sltmp, ctx->kinv, ctx->r,
                            ctx->ec);
    }
    if (ret <= 0)
        return 0;

    *siglen = sltmp;
    return 1;
}
```

The above function seems to be problematic. with
https://github.com/openssl/openssl/blob/d8fa4cf76308924daaf2335c6c0ff2f7334a5b26/providers/implementations/signature/ecdsa_sig.c#L177
returning different value than 
https://github.com/openssl/openssl/blob/d8fa4cf76308924daaf2335c6c0ff2f7334a5b26/providers/implementations/signature/ecdsa_sig.c#L204

Is this expected? If yes, what is a sure way to get the exact signature size without having to call `EVP_DigestSign` twice? If this is not expected, could you fix it please?

To give you more context, the issue originates while using PKCS11 API `C_Sign`. When the `pSignature` parameter is `NULL`, we need to return the actual size of the memory required for the signature.

![image](https://github.com/openssl/openssl/assets/29835993/204e9210-ba74-4ac8-99fb-371ccefbb2da)

So, to get the right signature , this means two calls to `EVP_DigestSign` (one will NULL and one with allocated memory for signature  returned from the first call) followed by two more `EVP_DigestSign` calls.

Further, this is also not guaranteed to work as, if we consider and allocate the actual length returned from the second call to `EVP_DigestSign`, it will fail in this check the next time we call `EVP_DigestSign` since ECDSA_Size in line 177 does not take the actual data to sign to consider for the signature length to be returned.

https://github.com/openssl/openssl/blob/d8fa4cf76308924daaf2335c6c0ff2f7334a5b26/providers/implementations/signature/ecdsa_sig.c#L192

Please let me know. Thanks.

**Comment by mattcaswell** [comment, marked as answer]
An ECDSA signature is non-deterministic in nature - you will get a different signature every time you run the signature process. This is a fundamental property of the ECDSA algorithm. An ECDSA signature is actually composed of two variable length integer values `r` and `s`. The output from `EVP_DigestSign` actually gives you the DER encoding of the ECDSA_SIG structure which contains those two `r` and `s` values. Given that the underlying `r` and `s` are variable length and will change each time you create the signature, the size of that signature may vary each time too. It is not possible to determine what the *actual* length of the signature will be until you create it. The best we can do is give an upper bound which we know all signatures will fit into.

So, to answer your questions:

> Is this expected?

For the reasons I give above - yes, this is expected.

> If yes, what is a sure way to get the exact signature size without having to call EVP_DigestSign twice? 

The only way to get an exact signature size is to actually create the signature. Up until that point the best we can do is give an upper bound.

**Comment by svenkata9** [comment]
Thanks @mattcaswell. Appreciate the quick response.
